src/main/java/de/taimos/pipeline/aws/cloudformation/CFNCreateChangeSetStep.java

```diff
@@ -0,0 +1,321 @@
+	public static class DescriptorImpl extends AbstractStepDescriptorImpl {
+		
+		public DescriptorImpl() {
+			super(Execution.class);
+		}
+		
+		@Override
+		public String getFunctionName() {
+			return "cfnCreateChangeSet";
+		}
+		
+		@Override
+		public String getDisplayName() {
+			return "Create CloudFormation change set";
+		}
+	}
+	
+	public static class Execution extends AbstractStepExecutionImpl {
+		
+		@Inject
+		private transient CFNCreateChangeSetStep step;
+		@StepContextParameter
+		private transient EnvVars envVars;
+		@StepContextParameter
+		private transient FilePath workspace;
+		@StepContextParameter
+		private transient TaskListener listener;
+		
+		@Override
+		public boolean start() throws Exception {
+			final String changeSet = this.step.getChangeSet();
+			final String stack = this.step.getStack();
+			final String file = this.step.getFile();
+			final String url = this.step.getUrl();
+			final Boolean create = this.step.getCreate();
+			final String roleArn = this.step.getRoleArn();
+
+			final Collection<Parameter> params= this.parseParamsFile(this.step.getParamsFile());
+			params.addAll(this.parseParams(this.step.getParams()));
+			
+			final Collection<Parameter> keepParams = this.parseKeepParams(this.step.getKeepParams());
+			final Collection<Tag> tags = this.parseTags(this.step.getTags());
+
+			Preconditions.checkArgument(changeSet != null && !changeSet.isEmpty(), "Change Set must not be null or empty");
+
+			Preconditions.checkArgument(stack != null && !stack.isEmpty(), "Stack must not be null or empty");
+
+			new Thread("cfnCreateChangeSet-" + changeSet) {
+				@Override
+				public void run() {
+					try {
+						AmazonCloudFormationClient client = AWSClientFactory.create(AmazonCloudFormationClient.class, Execution.this.envVars);
+						CloudFormationStack cfnStack = new CloudFormationStack(client, stack, Execution.this.listener);
+						if (cfnStack.exists()) {
+							Execution.this.listener.getLogger().format("Creating CloudFormation change set %s for existing stack %s %n", changeSet, stack);
+							ArrayList<Parameter> parameters = new ArrayList<>(params);
+							parameters.addAll(keepParams);
+							cfnStack.createChangeSet(changeSet, Execution.this.readTemplate(file), url, parameters, tags, Execution.this.step.getPollInterval(), ChangeSetType.UPDATE, roleArn);
+						} else if (create) {
+							Execution.this.listener.getLogger().format("Creating CloudFormation change set %s for new stack %s %n", changeSet, stack);
+							cfnStack.createChangeSet(changeSet, Execution.this.readTemplate(file), url, params, tags, Execution.this.step.getPollInterval(), ChangeSetType.CREATE, roleArn);
+						} else {
+							Execution.this.listener.getLogger().println("No stack found with the name and skipped change set creation due to configuration.");
+						}
+						Execution.this.listener.getLogger().println("Create change set complete");
+						Execution.this.getContext().onSuccess(cfnStack.describeOutputs());
```

Comment:
Does it make sense to return the outputs as the change was not executed yet?

Reply:
It does not explode because a temporary stack will be created by AWS (status review in progress I think), but I agree that it does not make much sense.

Looking at this more closely, I suspect that it is also a bug to try and return outputs if the creation is skipped altogether (missing stack  + create==false). I simply copied this code from the CFNUpdateStep, but I expect it to result in an Exception, as the describeOutputs method does not check for the existence of the stack.

Reply:
Great catch. I missed this when reviewing the "create" PullRequest. I will fix it in the cfnUpdate step.